### PR TITLE
fix(core): follow BS App Automate tmp dir relocation for Maestro screenshots

### PR DESCRIPTION
**Pull request by prklm10**
## Summary
- BrowserStack hosts relocated the App Automate session directories away from `/tmp/{sessionId}`, breaking the Maestro screenshot relay's hardcoded `/tmp` globs and scope roots — screenshot lookups started 404'ing.
- All BrowserStack-mode paths (glob patterns, manual-walker fallback, and the `scopeRoot` used by the realpath containment check) now route through a new `appAutomateTmpDir()` helper.
- The helper reads a new `PERCY_APP_AUTOMATE_TMP_DIR` env var — injected by the host with the relocated root — and falls back to the legacy `/tmp` convention when unset, so no internal infra path is hardcoded in the CLI and hosts with older injection keep working during rollout. Trailing separators are trimmed, backslashes normalized for the glob, and non-absolute values fall back to `/tmp`.
- Self-hosted mode is untouched — it continues to scope via `PERCY_MAESTRO_SCREENSHOT_DIR`.

## Test plan
- Existing `/percy/maestro-screenshot` specs (Android + iOS globs, `filePath` acceptance/containment, PNG-fill, regions) — all passing against the default root.
- New `PERCY_APP_AUTOMATE_TMP_DIR override` specs: globbing under an overridden root (real-fs fixture), trailing-slash tolerance, non-absolute fallback to `/tmp`, direct trim/fallback assertions on the exported helper, and `filePath` containment re-scoping to the overridden root.
- Full `packages/core/test/api.test.js` run locally: all maestro specs pass; the single unrelated failure (`when the server is disabled…`, ECONNREFUSED vs AggregateError) also fails on a clean master checkout — pre-existing local Node environment issue.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/core/src/maestro-screenshot-file.js [hunk 1]
```diff
@@ -2,6 +2,18 @@ import fs from 'fs';
...
 
+// Root under which BrowserStack App Automate hosts write Maestro session
+// artifacts (screenshots + debug output). Hosts relocated it from /tmp to
+// /usr/local/.browserstack/app-automate-tmp; PERCY_APP_AUTOMATE_TMP_DIR lets
+// the host injection move it again without a CLI release. Trailing separators
+// are trimmed so glob patterns and the realpath prefix check compose cleanly.
+// BrowserStack-mode only — self-hosted scoping stays on
+// PERCY_MAESTRO_SCREENSHOT_DIR.
+export function appAutomateTmpDir() {
+  let dir = process.env.PERCY_APP_AUTOMATE_TMP_DIR || '/usr/local/.browserstack/app-automate-tmp';
```
Comment (prklm10) [inline packages/core/src/maestro-screenshot-file.js]: **[Medium] Rollout-skew risk: no legacy `/tmp` fallback**

If any BS host fleet segment (Linux/Android vs macOS/iOS are separate fleets) still writes to `/tmp/{sid}`, this CLI version globs only the new root and hard-404s — and non-relocated hosts by definition run older injection that won't set `PERCY_APP_AUTOMATE_TMP_DIR=/tmp`.

**Suggestion:** Confirm the relocation is fleet-wide for both Android and iOS hosts before release (stragglers can inject `PERCY_APP_AUTOMATE_TMP_DIR=/tmp`), or add a legacy fallback that retries the `/tmp` pattern (with matching scopeRoot) when the new-root glob is empty and the env var is unset.

_Reviewer: stack:code-reviewer_

File: packages/core/src/maestro-screenshot-file.js [hunk 2]
```diff
@@ -2,6 +2,18 @@ import fs from 'fs';
...
+  return dir.replace(/[/\\]+$/, '');
```
Comment (prklm10) [inline packages/core/src/maestro-screenshot-file.js]: **[Low] No shape validation on `PERCY_APP_AUTOMATE_TMP_DIR`**

Asymmetric with `PERCY_MAESTRO_SCREENSHOT_DIR`, which gets absolute-path + existing-dir checks with actionable 400s. A relative value here yields a cwd-relative glob surfacing only as an opaque 404; `/` trims to `''`, silently making the filesystem root the tmp root. Not a security issue (env is host-injected; containment fails closed), but host misconfiguration will be painful to debug.

**Suggestion:** After trimming, fall back to the default (optionally with a warn log) when the value is empty or not absolute.

_Reviewer: stack:code-reviewer_

File: packages/core/src/maestro-screenshot.js [hunk 3]
```diff
@@ -125,8 +127,8 @@ export async function handleMaestroScreenshot(req, res, percy) {
     scopeRoot = dir;
   } else {
     scopeRoot = platform === 'ios'
-      ? `/tmp/${sessionId}`
-      : `/tmp/${sessionId}_test_suite`;
+      ? `${appAutomateTmpDir()}/${sessionId}`
```
Comment (prklm10) [inline packages/core/src/maestro-screenshot.js:129]: **[Low] scopeRoot and glob root resolved by separate env reads**

The containment boundary (scopeRoot) and the glob search root come from independent `appAutomateTmpDir()` calls; a mid-request env change would make them diverge. Divergence fails closed (404), so this is coherence-only.

**Suggestion:** Optionally resolve the root once in the handler and derive the glob root from the already-passed `scopeRoot`, as the self-hosted branch does.

_Reviewer: stack:code-reviewer_

File: packages/core/test/api.test.js [hunk 4]
```diff
@@ -1826,6 +1829,68 @@ describe('API Server', () => {
       expect(payload.tiles[0].content).toBe(Buffer.from('PNGBYTES-ANDROID').toString('base64'));
     });
 
+    // PERCY_APP_AUTOMATE_TMP_DIR override — BS hosts relocated the session
+    // tmp root (previously /tmp, now /usr/local/.browserstack/app-automate-tmp);
+    // the env var lets host injection move it again without a CLI release.
+    // Real-fs root (matched by the top-level $bypass) — fast-glob caches its
+    // fs bindings on first import, so a memfs-only root created in a later
+    // test is invisible to it; real-fs fixtures sidestep the staleness.
+    describe('PERCY_APP_AUTOMATE_TMP_DIR override', () => {
```
Comment (prklm10) [inline packages/core/test/api.test.js:1840]: **[Low] Override tests cover the Android branch only**

The iOS glob and scopeRoot expressions are never exercised under an override (low risk — same `tmpRoot` expression as Android).

**Suggestion:** Optionally add one iOS spec mirroring the Android override test, with a fixture under `${CUSTOM_ROOT}/${SID}/emu_maestro_debug_abc/...`.

_Reviewer: stack:code-reviewer_

File: packages/core/test/api.test.js [hunk 5]
```diff
@@ -1826,6 +1829,68 @@ describe('API Server', () => {
...
+        await percy.start();
+
+        await expectAsync(postMaestro({ name: SS_NAME, sessionId: SID, platform: 'android' }))
+          .toBeResolvedTo(jasmine.objectContaining({ success: true }));
+
+        let [payload] = percy.upload.calls.mostRecent().args;
+        expect(payload.tiles[0].content).toBe(Buffer.from('PNGBYTES-CUSTOM-ROOT').toString('base64'));
+      });
+
+      it('tolerates a trailing slash on the override', async () => {
```
Comment (prklm10) [inline packages/core/test/api.test.js:1870]: **[Low] Trailing-slash spec doesn't strictly pin the trim behavior**

fast-glob normalizes `//` in patterns and realpath collapses duplicate separators, so this spec would likely pass even without the `replace(/[/\\]+$/, '')` trim — it documents intent more than it guards the code.

**Suggestion:** Optionally add a direct unit assertion on the exported helper, e.g. `expect(appAutomateTmpDir()).toBe(CUSTOM_ROOT)` with the env set to `${CUSTOM_ROOT}///`.

_Reviewer: stack:code-reviewer_

**Review by prklm10 — COMMENTED** [review]
Claude Code Review (automated) — 5 inline finding(s). Full report in the PR comment below. Verdict: Passed.

**Comment by prklm10** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/cli/pull/2353  •  **Head:** f28af57  •  **Reviewers:** stack:code-reviewer

## Summary
Fixes the Maestro screenshot relay after BrowserStack App Automate hosts relocated session artifact directories from `/tmp/{sessionId}` to the relocated App Automate session root: all BS-mode paths (glob patterns, manual-walker fallback, containment scope roots) now route through a new `appAutomateTmpDir()` helper reading `PERCY_APP_AUTOMATE_TMP_DIR` (default `the relocated root`); self-hosted mode is unchanged.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | Only a path constant; env var is host-injected, never request-controlled |
| High | Security | Authentication/authorization checks present | N/A | Local relay endpoint; no auth surface changed |
| High | Security | Input validation and sanitization | Pass | SAFE_ID validation on `name`/`sessionId` untouched; realpath + scope-root containment invariant preserved and fail-closed |
| High | Security | No IDOR — resource ownership validated | Pass | Cross-session containment re-verified under the overridden root by a new test |
| High | Security | No SQL injection (parameterized queries) | N/A | No database code |
| High | Correctness | Logic is correct, handles edge cases | Pass | Trailing-separator trim, backslash normalization for glob; grep-verified no BS-mode `/tmp` path survives in source |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Missing screenshots still surface as actionable 404s with the searched pattern |
| High | Correctness | No race conditions or concurrency issues | Pass | scopeRoot and glob root come from separate env reads (Finding 3), but divergence fails closed (404) |
| Medium | Testing | New code has corresponding tests | Pass | 3 new override specs (custom root, trailing slash, containment re-scoping); existing specs migrated to the new default root |
| Medium | Testing | Error paths and edge cases tested | Pass | Containment/404 paths covered; iOS override branch untested (Finding 4, low risk — same expression as Android) |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Full api.test.js run: 126/127; the 1 failure is a pre-existing environment flake reproduced on clean master |
| Medium | Performance | No N+1 queries or unbounded data fetching | N/A | No data-fetching changes |
| Medium | Performance | Long-running tasks use background jobs | N/A | Not applicable |
| Medium | Quality | Follows existing codebase patterns | Pass | Mirrors PERCY_MAESTRO_SCREENSHOT_DIR env-var pattern and the existing real-fs `$bypass` test convention |
| Medium | Quality | Changes are focused (single concern) | Pass | 3 files, all scoped to the tmp-root relocation |
| Low | Quality | Meaningful names, no dead code | Pass | `appAutomateTmpDir()` is descriptive; no dead code introduced |
| Low | Quality | Comments explain why, not what | Pass | Comments updated everywhere paths changed, incl. the macOS-symlink rationale generalized rather than deleted |
| Low | Quality | No unnecessary dependencies added | Pass | No dependency changes |

## Findings

- **File:** `packages/core/src/maestro-screenshot-file.js:13`
- **Severity:** Medium
- **Reviewer:** stack:code-reviewer
- **Issue:** Rollout-skew risk: no legacy `/tmp` fallback. If any BS host fleet segment (Linux/Android vs macOS/iOS are different fleets) still writes to `/tmp/{sid}`, this CLI version globs only the new root and hard-404s; non-relocated hosts by definition run older injection that won't set `PERCY_APP_AUTOMATE_TMP_DIR=/tmp`.
- **Suggestion:** Confirm the relocation is fleet-wide for both Android and iOS hosts before release (and note the `PERCY_APP_AUTOMATE_TMP_DIR=/tmp` escape hatch for stragglers), or add a legacy fallback that retries the `/tmp` pattern (with matching scopeRoot) when the new-root glob is empty and the env var is unset.

- **File:** `packages/core/src/maestro-screenshot-file.js:14`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** No shape validation on `PERCY_APP_AUTOMATE_TMP_DIR`, asymmetric with `PERCY_MAESTRO_SCREENSHOT_DIR` (which gets absolute-path + existing-dir checks with actionable 400s). A relative value yields a cwd-relative glob surfacing only as an opaque 404; `/` trims to `''`, silently making the filesystem root the tmp root. Not a security issue (env is host-injected; containment fails closed), but host misconfiguration will be painful to debug.
- **Suggestion:** Validate in `appAutomateTmpDir()`: after trimming, fall back to the default (optionally with a warn log) when the value is empty or not absolute.

- **File:** `packages/core/src/maestro-screenshot.js:130`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** scopeRoot (containment boundary) and the glob search root are computed by independent `appAutomateTmpDir()` env reads; a mid-request env change would make them diverge. Divergence fails closed (404), so this is coherence-only.
- **Suggestion:** Optionally resolve the root once in the handler and derive the glob root from the already-passed `scopeRoot`, as the self-hosted branch does.

- **File:** `packages/core/test/api.test.js:1838`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** Override tests cover the Android branch only; the iOS glob and scopeRoot expressions are never exercised under an override (low risk — same `tmpRoot` expression as Android).
- **Suggestion:** Optionally add one iOS spec mirroring the Android override test, with a fixture under `${CUSTOM_ROOT}/${SID}/emu_maestro_debug_abc/...`.

- **File:** `packages/core/test/api.test.js:1868`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** The trailing-slash spec would likely pass even without the trim (fast-glob normalizes `//` and realpath collapses duplicate separators) — it documents intent more than it guards the code.
- **Suggestion:** Optionally add a direct unit assertion on the exported helper, e.g. `expect(appAutomateTmpDir()).toBe(CUSTOM_ROOT)` with the env set to `${CUSTOM_ROOT}///`.

Informational (not counted against the verdict): the fixed real-fs fixture name `percy-bs-tmp-real-root` in `os.tmpdir()` could collide across concurrent suite runs on one machine (consistent with the pre-existing `percy-self-hosted-real` convention); one unrelated pre-existing flake (`page discovery … captures requests from workers`) was observed during the local full-suite run and is not attributable to this change.

---

**Verdict: PASS** — approved by stack:code-reviewer; 1 Medium deployment-verification question (fleet-wide relocation confirmation) and 4 non-blocking polish items.

File: packages/core/src/maestro-screenshot-file.js [hunk 6]
```diff
@@ -2,6 +2,19 @@ import fs from 'fs';
...
+// Root under which BrowserStack App Automate hosts write Maestro session
+// artifacts (screenshots + debug output). Historically /tmp; hosts that
+// relocate it inject PERCY_APP_AUTOMATE_TMP_DIR with the new root, so the
+// location is never hardcoded here and can move again without a CLI release.
+// Trailing separators are trimmed so glob patterns and the realpath prefix
+// check compose cleanly; a non-absolute value falls back to /tmp rather than
+// producing a cwd-relative glob. BrowserStack-mode only — self-hosted scoping
+// stays on PERCY_MAESTRO_SCREENSHOT_DIR.
+export function appAutomateTmpDir() {
+  let dir = (process.env.PERCY_APP_AUTOMATE_TMP_DIR || '/tmp').replace(/[/\\]+$/, '');
```
Comment (prklm10) [inline packages/core/src/maestro-screenshot-file.js:14]: **[Low] Silent fallback hides host misconfiguration (+ no whitespace trim)**

A set-but-invalid value — non-absolute, trimming to empty, or carrying incidental whitespace from shell injection (whitespace fails `isAbsolute` too) — is silently discarded; the operator only sees `Screenshot not found` 404s against `/tmp`. Partially mitigated because the 404 message echoes the searched pattern.

**Suggestion:** two-line hardening: add `.trim()` before the separator trim, and emit a debug log when a set value is rejected, e.g. `Ignoring non-absolute PERCY_APP_AUTOMATE_TMP_DIR: <raw>`.

_Reviewer: stack:code-reviewer_

File: packages/core/src/maestro-screenshot.js [hunk 7]
```diff
@@ -125,8 +126,8 @@ export async function handleMaestroScreenshot(req, res, percy) {
...
```
Comment (prklm10) [inline packages/core/src/maestro-screenshot.js:129]: **[Low] Tmp root still read independently at three sites per request**

scopeRoot here, the glob root, and the manual walker each call `appAutomateTmpDir()` separately. Divergence between reads fails closed (404, never an escape), so this is coherence-only — carried over from the prior round.

**Suggestion:** in the BS branch of `locateScreenshot`, derive the glob from the already-passed `scopeRoot` (`{scopeRoot}/logs/*/screenshots/…` Android, `{scopeRoot}/*_maestro_debug_*/**` iOS), which also unifies the two branches' slash normalization. Fine as a fast-follow.

_Reviewer: stack:code-reviewer_

File: packages/core/test/api.test.js [hunk 8]
```diff
@@ -1826,6 +1831,90 @@ describe('API Server', () => {
       expect(payload.tiles[0].content).toBe(Buffer.from('PNGBYTES-ANDROID').toString('base64'));
     });
 
+    // PERCY_APP_AUTOMATE_TMP_DIR override — BS hosts that relocate the
+    // session tmp root away from /tmp inject the new root via this env var,
+    // so the location is never hardcoded in the CLI.
+    // Real-fs root (matched by the top-level $bypass) — fast-glob caches its
+    // fs bindings on first import, so a memfs-only root created in a later
+    // test is invisible to it; real-fs fixtures sidestep the staleness.
+    describe('PERCY_APP_AUTOMATE_TMP_DIR override', () => {
```
Comment (prklm10) [inline packages/core/test/api.test.js:1840]: **[Low] Override coverage is Android-glob-only end-to-end**

The iOS override glob shares the helper (low risk, and helper semantics are now pinned platform-independently by the direct assertions below) but is never exercised under a custom root.

**Suggestion:** optionally add one iOS spec under `CUSTOM_ROOT` reusing the existing `makePngHeader` fixture helper.

_Reviewer: stack:code-reviewer_

**Review by prklm10 — COMMENTED** [review]
Claude Code Review (automated) — 3 inline finding(s). Full report in the PR comment below. Verdict: Passed.

**Comment by prklm10** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/cli/pull/2353  •  **Head:** d07e89a  •  **Reviewers:** stack:code-reviewer

## Summary
Fixes the Maestro screenshot relay after BrowserStack App Automate hosts relocated session artifact directories away from `/tmp/{sessionId}`: all BS-mode paths (glob patterns, manual-walker fallback, containment scope roots) route through a new `appAutomateTmpDir()` helper that reads `PERCY_APP_AUTOMATE_TMP_DIR` (host-injected relocated root — no infra path hardcoded in the public CLI) and falls back to the legacy `/tmp` convention when unset or non-absolute; self-hosted mode is unchanged.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | No infra path or secret in source; env var is host-injected, never request-controlled |
| High | Security | Authentication/authorization checks present | N/A | Local relay endpoint; no auth surface changed |
| High | Security | Input validation and sanitization | Pass | SAFE_ID validation untouched; env value gets absolute-path check with safe `/tmp` fallback |
| High | Security | No IDOR — resource ownership validated | Pass | realpath + prefix containment preserved; spec pins that containment scopes to whichever root is active |
| High | Security | No SQL injection (parameterized queries) | N/A | No database code |
| High | Correctness | Logic is correct, handles edge cases | Pass | Trailing-separator trim, non-absolute → `/tmp` fallback, backslash normalization; grep-verified no BS-mode `/tmp` literal survives outside comments |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | 404s echo the searched pattern; silent-fallback observability noted as Low (Finding 1) |
| High | Correctness | No race conditions or concurrency issues | Pass | Root read at multiple sites per request (Finding 2) but divergence fails closed (404), never an escape |
| Medium | Testing | New code has corresponding tests | Pass | 5 override specs incl. direct trim/fallback helper assertions and non-absolute fallback behavior |
| Medium | Testing | Error paths and edge cases tested | Pass | Fallback + containment paths pinned; iOS override glob still not exercised end-to-end (Finding 3, low risk) |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Full maestro-screenshot suite (66 specs) verified passing on this head |
| Medium | Performance | No N+1 queries or unbounded data fetching | N/A | No data-fetching changes |
| Medium | Performance | Long-running tasks use background jobs | N/A | Not applicable |
| Medium | Quality | Follows existing codebase patterns | Pass | Mirrors PERCY_MAESTRO_SCREENSHOT_DIR env pattern and the real-fs `$bypass` test convention |
| Medium | Quality | Changes are focused (single concern) | Pass | 3 files, all scoped to the tmp-root relocation |
| Low | Quality | Meaningful names, no dead code | Pass | Single exported helper; no dead code |
| Low | Quality | Comments explain why, not what | Pass | Comments explain injection model, trim rationale, and fallback semantics without exposing infra paths |
| Low | Quality | No unnecessary dependencies added | Pass | No dependency changes |

Prior-round concerns: rollout-skew (legacy fallback) and env shape validation — **resolved with test pins**; trailing-slash trim — **resolved via direct helper assertions**; separate env reads and iOS override coverage — carried forward as Low (Findings 2–3).

## Findings

- **File:** `packages/core/src/maestro-screenshot-file.js:14`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** Silent fallback hides host misconfiguration: a set-but-invalid value (non-absolute, trims to empty, or carrying incidental whitespace from shell injection — whitespace fails `isAbsolute` too) is discarded without any signal; the operator only sees `Screenshot not found` 404s against `/tmp`. Partially mitigated because the 404 message echoes the searched pattern.
- **Suggestion:** Two-line hardening in the helper: add `.trim()` before the separator trim, and emit a debug log when a set value is rejected (e.g. `Ignoring non-absolute PERCY_APP_AUTOMATE_TMP_DIR: <raw>`).

- **File:** `packages/core/src/maestro-screenshot.js:129`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** The tmp root is still read from the env independently at three sites per request (scopeRoot here; glob root and manual walker in `maestro-screenshot-file.js`). Divergence between reads fails closed (404, never an escape), so this is coherence-only — but it is the same structural concern raised last round.
- **Suggestion:** In the BS branch of `locateScreenshot`, derive the glob from the already-passed `scopeRoot` (`{scopeRoot}/logs/*/screenshots/…` Android, `{scopeRoot}/*_maestro_debug_*/**` iOS) — this also unifies the two branches' slash normalization. Fine as a fast-follow.

- **File:** `packages/core/test/api.test.js:1840`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** Override coverage is Android-glob-only end-to-end; the iOS override glob shares the helper (risk is low, and helper semantics are now pinned platform-independently) but is never exercised under a custom root.
- **Suggestion:** Optionally add one iOS spec under `CUSTOM_ROOT` reusing the existing `makePngHeader` fixture helper.

Informational (not counted against the verdict): on win32 a degenerate root like `C:\` trims to a non-absolute value and falls back to `/tmp` (drive-relative there) — irrelevant in production since BS hosts are POSIX and CI specs use `os.tmpdir()` paths that survive the trim.

---

**Verdict: PASS** — approved by stack:code-reviewer; both prior blocking concerns resolved with test pins, containment security preserved under relocated roots; remaining findings are Low polish (a two-line helper hardening is recommended pre-merge, the rest can follow up).

**Comment by prklm10** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/cli/pull/2353  •  **Head:** fb9afa0  •  **Reviewers:** stack:code-reviewer

## Summary
Fixes the Maestro screenshot relay after BrowserStack App Automate hosts relocated session artifact directories away from `/tmp/{sessionId}`: all BS-mode paths route through a new `appAutomateTmpDir()` helper reading `PERCY_APP_AUTOMATE_TMP_DIR` (host-injected relocated root — no infra path hardcoded in the public CLI) with a legacy `/tmp` fallback when unset or non-absolute; self-hosted mode is unchanged.

> **Note:** Head `fb9afa0` is a history-only squash — its tree and PR diff are byte-identical to the previously reviewed head `d07e89a` (verified via `git diff --quiet` and a byte-compare of the `gh pr diff` snapshots). The `stack:code-reviewer` findings below are from that review of this exact diff. The three inline comments from review [#4786463906](https://github.com/percy/cli/pull/2353#pullrequestreview-4786463906) anchor to the same unchanged lines and remain valid, so they are not re-posted.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | No infra path or secret in source; env var is host-injected, never request-controlled |
| High | Security | Authentication/authorization checks present | N/A | Local relay endpoint; no auth surface changed |
| High | Security | Input validation and sanitization | Pass | SAFE_ID validation untouched; env value gets absolute-path check with safe `/tmp` fallback |
| High | Security | No IDOR — resource ownership validated | Pass | realpath + prefix containment preserved; spec pins that containment scopes to whichever root is active |
| High | Security | No SQL injection (parameterized queries) | N/A | No database code |
| High | Correctness | Logic is correct, handles edge cases | Pass | Trailing-separator trim, non-absolute → `/tmp` fallback, backslash normalization; no BS-mode `/tmp` literal survives outside comments |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | 404s echo the searched pattern; silent-fallback observability noted as Low (Finding 1) |
| High | Correctness | No race conditions or concurrency issues | Pass | Root read at multiple sites per request (Finding 2) but divergence fails closed (404), never an escape |
| Medium | Testing | New code has corresponding tests | Pass | 5 override specs incl. direct trim/fallback helper assertions and non-absolute fallback behavior |
| Medium | Testing | Error paths and edge cases tested | Pass | Fallback + containment paths pinned; iOS override glob still not exercised end-to-end (Finding 3, low risk) |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Full maestro-screenshot suite (66 specs) verified passing on this exact tree |
| Medium | Performance | No N+1 queries or unbounded data fetching | N/A | No data-fetching changes |
| Medium | Performance | Long-running tasks use background jobs | N/A | Not applicable |
| Medium | Quality | Follows existing codebase patterns | Pass | Mirrors PERCY_MAESTRO_SCREENSHOT_DIR env pattern and the real-fs `$bypass` test convention |
| Medium | Quality | Changes are focused (single concern) | Pass | 3 files, single squashed commit, all scoped to the tmp-root relocation |
| Low | Quality | Meaningful names, no dead code | Pass | Single exported helper; no dead code |
| Low | Quality | Comments explain why, not what | Pass | Comments explain injection model, trim rationale, and fallback semantics without exposing infra paths |
| Low | Quality | No unnecessary dependencies added | Pass | No dependency changes |

## Findings

- **File:** `packages/core/src/maestro-screenshot-file.js:14`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** Silent fallback hides host misconfiguration: a set-but-invalid value (non-absolute, trims to empty, or carrying incidental whitespace — whitespace fails `isAbsolute` too) is discarded without any signal; the operator only sees 404s against `/tmp`. Partially mitigated because the 404 message echoes the searched pattern.
- **Suggestion:** Two-line hardening: add `.trim()` before the separator trim, and emit a debug log when a set value is rejected.

- **File:** `packages/core/src/maestro-screenshot.js:129`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** The tmp root is read from the env independently at three sites per request (scopeRoot, glob root, manual walker). Divergence fails closed (404, never an escape) — coherence-only.
- **Suggestion:** Derive the glob from the already-passed `scopeRoot` in the BS branch of `locateScreenshot`; fine as a fast-follow.

- **File:** `packages/core/test/api.test.js:1840`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** Override coverage is Android-glob-only end-to-end; the iOS override glob shares the helper (low risk, semantics pinned platform-independently) but is never exercised under a custom root.
- **Suggestion:** Optionally add one iOS spec under `CUSTOM_ROOT` reusing the existing `makePngHeader` fixture helper.

Informational: on win32 a degenerate root like `C:\` trims to a non-absolute value and falls back to `/tmp` — irrelevant in production since BS hosts are POSIX and CI specs use `os.tmpdir()` paths that survive the trim.

---

**Verdict: PASS** — approved by stack:code-reviewer on this exact diff; remaining findings are Low polish already tracked as inline comments on the PR.

**Review by ninadbstack — APPROVED** [review]
